### PR TITLE
Add ShellCheck workflow

### PR DESCRIPTION
**Pull request by jamacku**

### CODE REVIEW

**Review by phracek — COMMENTED** [review]
The pull request itself LGTM. Thanks. @zmiklank Does it make sense to allow it in all our repositories?

**Comment by jamacku** [comment]
Just to be clear, this check will fail only on new defects.

**Comment by zmiklank** [comment]
@phracek what about adding this to our meeting agenda for Wednesday?

**Comment by zmiklank** [comment]
We discussed this proposal on the container meeting. The conclusion is to do the shell checking properly - first the repository has to be manually scanned for shell issues. If some issue is found it has to be either fixed or muted.
The differential shellcheck action can be added just after these steps.

**Comment by jamacku** [comment]
Here is some analysis that might help you when considering this PR.

I have merged the change proposed in this PR to the master branch in ma fork. Here is the output from workflow: https://github.com/jamacku/nginx-container/actions/runs/5759812410/attempts/1

Differential ShellCheck found the following ShellCheck defects with severity warning and higher (in total **82**):

```sh
shellcheck \                                                                                                                    
  --external-sources \
  --severity=warning \
  $( \
    grep -r -E '(^\s*((#|!)|(#\s*!)|(!\s*#))\s*(/usr(/local)?)?/bin/(env\s+)?(sh|ash|bash|dash|ksh|bats)\b)|(^\s*#\s*shellcheck\s+shell=(sh|ash|bash|dash|ksh|bats)\s*)|(^\s*#\s+-\*-\s+(sh|ash|bash|dash|ksh|bats)\s+-\*-\s*)|(^\s*#\s+vim?:\s+(set\s+)?(ft|filetype)=(sh|ash|bash|dash|ksh|bats)\s*)' './' \
    | cut -d: -f1 \
  ) ./1.24/test/start-hook-test-app/nginx-start/init.sh ./1.22/test/start-hook-test-app/nginx-start/init.sh  ./1.20/test/start-hook-test-app/nginx-start/init.sh          

In ./1.20/root/usr/share/container-scripts/nginx/common.sh line 4:
function get_matched_files() {
^-- SC2112 (warning): 'function' keyword is non-standard. Delete it.


In ./1.20/root/usr/share/container-scripts/nginx/common.sh line 5:
  local custom_dir default_dir
  ^--------------------------^ SC3043 (warning): In POSIX sh, 'local' is undefined.


In ./1.20/root/usr/share/container-scripts/nginx/common.sh line 16:
function process_extending_files() {
^-- SC2112 (warning): 'function' keyword is non-standard. Delete it.


In ./1.20/root/usr/share/container-scripts/nginx/common.sh line 17:
  local custom_dir default_dir
  ^--------------------------^ SC3043 (warning): In POSIX sh, 'local' is undefined.


In ./1.20/root/usr/share/container-scripts/nginx/common.sh line 25:
        source $custom_dir/$filename
        ^--------------------------^ SC3046 (warning): In POSIX sh, 'source' in place of '.' is undefined.
               ^-------------------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In ./1.20/root/usr/share/container-scripts/nginx/common.sh line 27:
        source $default_dir/$filename
        ^---------------------------^ SC3046 (warning): In POSIX sh, 'source' in place of '.' is undefined.
               ^--------------------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In ./1.20/root/usr/share/container-scripts/nginx/common.sh line 30:
  done <<<"$(get_matched_files "$custom_dir" "$default_dir" '*.sh' | sort -u)"
       ^-^ SC3011 (warning): In POSIX sh, here-strings are undefined.


In ./1.20/s2i/bin/run line 11:
if [ ! -v NGINX_LOG_TO_VOLUME -a -v NGINX_LOG_PATH ]; then
                              ^-- SC2166 (warning): Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In ./test/run-openshift-local-cluster line 10:
test_dir="$(readlink -f $(dirname ${BASH_SOURCE[0]}))"
                        ^--------------------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run-openshift-local-cluster line 60:
OS_TESTSUITE_RESULT=0
^-----------------^ SC2034 (warning): OS_TESTSUITE_RESULT appears unused. Verify use (or export if used externally).


In ./test/run line 14:
test_dir="$(readlink -f $(dirname "${BASH_SOURCE[0]}"))"
                        ^----------------------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 40:
  echo -e "\n\e[1m[INFO] $@...\e[0m\n"
                         ^-- SC2145 (error): Argument mixes string and array. Use * or separate argument.


In ./test/run line 48:
  image_exists $(cat $cid_file)
               ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 52:
  docker inspect --format="{{ .NetworkSettings.IPAddress }}" $(cat $cid_file)
                                                             ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 71:
  pushd ${test_dir}/${1} >/dev/null
  ^-- SC2164 (warning): Use 'pushd ... || exit' or 'pushd ... || return' in case pushd fails.

Did you mean: 
  pushd ${test_dir}/${1} >/dev/null || exit


In ./test/run line 75:
  popd >/dev/null
  ^-------------^ SC2164 (warning): Use 'popd ... || exit' or 'popd ... || return' in case popd fails.

Did you mean: 
  popd >/dev/null || exit


In ./test/run line 88:
      docker stop $(cat $cid_file)
                  ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 89:
      docker rm $(cat $cid_file)
                ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 99:
    TESTCASE_RESULT=1
    ^-------------^ SC2034 (warning): TESTCASE_RESULT appears unused. Verify use (or export if used externally).


In ./test/run line 134:
    echo "ERROR[/bin/bash -c "${run_cmd}"] Expected '${expected}', got '${out}'"
                              ^--------^ SC2027 (warning): The surrounding quotes actually unquote this. Remove or escape them.


In ./test/run line 151:
    echo "ERROR[exec /bin/bash -c "${run_cmd}"] Expected '${expected}', got '${out}'"
                                   ^--------^ SC2027 (warning): The surrounding quotes actually unquote this. Remove or escape them.


In ./test/run line 156:
    echo "ERROR[exec /bin/sh -ic "${run_cmd}"] Expected '${expected}', got '${out}'"
                                  ^--------^ SC2027 (warning): The surrounding quotes actually unquote this. Remove or escape them.


In ./test/run line 172:
      out=$(docker logs $(cat ${cid_file}) 2>&1)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 174:
      out=$(docker logs $(cat ${cid_file}) 2>/dev/null)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 176:
      out=$(docker logs $(cat ${cid_file}) 2>&1 1>/dev/null)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 179:
      out=$(docker exec $(cat ${cid_file}) cat ${logfile})
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 181:
      out=$(docker logs $(cat ${cid_file}) 2>&1)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 355:
  local perl_version=$(docker exec $(cat ${cid_file}) /bin/bash -c "perl -e 'print \"$^V\"'" 2>&1)
        ^----------^ SC2155 (warning): Declare and assign separately to avoid masking return values.
                                   ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./test/run line 416:
  local dockerfile_tmp=$(mktemp)
        ^------------^ SC2155 (warning): Declare and assign separately to avoid masking return values.


In ./test/run-openshift-remote-cluster line 37:
TEST_SUMMARY=''
^----------^ SC2034 (warning): TEST_SUMMARY appears unused. Verify use (or export if used externally).


In ./1.22/root/usr/share/container-scripts/nginx/common.sh line 4:
function get_matched_files() {
^-- SC2112 (warning): 'function' keyword is non-standard. Delete it.


In ./1.22/root/usr/share/container-scripts/nginx/common.sh line 5:
  local custom_dir default_dir
  ^--------------------------^ SC3043 (warning): In POSIX sh, 'local' is undefined.


In ./1.22/root/usr/share/container-scripts/nginx/common.sh line 16:
function process_extending_files() {
^-- SC2112 (warning): 'function' keyword is non-standard. Delete it.


In ./1.22/root/usr/share/container-scripts/nginx/common.sh line 17:
  local custom_dir default_dir
  ^--------------------------^ SC3043 (warning): In POSIX sh, 'local' is undefined.


In ./1.22/root/usr/share/container-scripts/nginx/common.sh line 25:
        source $custom_dir/$filename
        ^--------------------------^ SC3046 (warning): In POSIX sh, 'source' in place of '.' is undefined.
               ^-------------------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In ./1.22/root/usr/share/container-scripts/nginx/common.sh line 27:
        source $default_dir/$filename
        ^---------------------------^ SC3046 (warning): In POSIX sh, 'source' in place of '.' is undefined.
               ^--------------------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In ./1.22/root/usr/share/container-scripts/nginx/common.sh line 30:
  done <<<"$(get_matched_files "$custom_dir" "$default_dir" '*.sh' | sort -u)"
       ^-^ SC3011 (warning): In POSIX sh, here-strings are undefined.


In ./1.22/s2i/bin/run line 11:
if [ ! -v NGINX_LOG_TO_VOLUME -a -v NGINX_LOG_PATH ]; then
                              ^-- SC2166 (warning): Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In ./1.22/test/run line 14:
test_dir="$(readlink -f $(dirname "${BASH_SOURCE[0]}"))"
                        ^----------------------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 40:
  echo -e "\n\e[1m[INFO] $@...\e[0m\n"
                         ^-- SC2145 (error): Argument mixes string and array. Use * or separate argument.


In ./1.22/test/run line 48:
  image_exists $(cat $cid_file)
               ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 52:
  docker inspect --format="{{ .NetworkSettings.IPAddress }}" $(cat $cid_file)
                                                             ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 71:
  pushd ${test_dir}/${1} >/dev/null
  ^-- SC2164 (warning): Use 'pushd ... || exit' or 'pushd ... || return' in case pushd fails.

Did you mean: 
  pushd ${test_dir}/${1} >/dev/null || exit


In ./1.22/test/run line 75:
  popd >/dev/null
  ^-------------^ SC2164 (warning): Use 'popd ... || exit' or 'popd ... || return' in case popd fails.

Did you mean: 
  popd >/dev/null || exit


In ./1.22/test/run line 88:
      docker stop $(cat $cid_file)
                  ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 89:
      docker rm $(cat $cid_file)
                ^--------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 99:
    TESTCASE_RESULT=1
    ^-------------^ SC2034 (warning): TESTCASE_RESULT appears unused. Verify use (or export if used externally).


In ./1.22/test/run line 134:
    echo "ERROR[/bin/bash -c "${run_cmd}"] Expected '${expected}', got '${out}'"
                              ^--------^ SC2027 (warning): The surrounding quotes actually unquote this. Remove or escape them.


In ./1.22/test/run line 151:
    echo "ERROR[exec /bin/bash -c "${run_cmd}"] Expected '${expected}', got '${out}'"
                                   ^--------^ SC2027 (warning): The surrounding quotes actually unquote this. Remove or escape them.


In ./1.22/test/run line 156:
    echo "ERROR[exec /bin/sh -ic "${run_cmd}"] Expected '${expected}', got '${out}'"
                                  ^--------^ SC2027 (warning): The surrounding quotes actually unquote this. Remove or escape them.


In ./1.22/test/run line 172:
      out=$(docker logs $(cat ${cid_file}) 2>&1)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 174:
      out=$(docker logs $(cat ${cid_file}) 2>/dev/null)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 176:
      out=$(docker logs $(cat ${cid_file}) 2>&1 1>/dev/null)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 179:
      out=$(docker exec $(cat ${cid_file}) cat ${logfile})
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 181:
      out=$(docker logs $(cat ${cid_file}) 2>&1)
                        ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 355:
  local perl_version=$(docker exec $(cat ${cid_file}) /bin/bash -c "perl -e 'print \"$^V\"'" 2>&1)
        ^----------^ SC2155 (warning): Declare and assign separately to avoid masking return values.
                                   ^----------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run line 416:
  local dockerfile_tmp=$(mktemp)
        ^------------^ SC2155 (warning): Declare and assign separately to avoid masking return values.


In ./1.22/test/run-openshift line 10:
test_dir="$(readlink -f $(dirname ${BASH_SOURCE[0]}))"
                        ^--------------------------^ SC2046 (warning): Quote this to prevent word splitting.


In ./1.22/test/run-openshift line 60:
OS_TESTSUITE_RESULT=0
^-----------------^ SC2034 (warning): OS_TESTSUITE_RESULT appears unused. Verify use (or export if used externally).


In ./1.22/test/run-openshift-remote-cluster line 37:
TEST_SUMMARY=''
^----------^ SC2034 (warning): TEST_SUMMARY appears unused. Verify use (or export if used externally).


In ./1.24/root/usr/share/container-scripts/nginx/common.sh line 4:
function get_matched_files() {
^-- SC2112 (warning): 'function' keyword is non-standard. Delete it.


In ./1.24/root/usr/share/container-scripts/nginx/common.sh line 5:
  local custom_dir default_dir
  ^--------------------------^ SC3043 (warning): In POSIX sh, 'local' is undefined.


In ./1.24/root/usr/share/container-scripts/nginx/common.sh line 16:
function process_extending_files() {
^-- SC2112 (warning): 'function' keyword is non-standard. Delete it.


In ./1.24/root/usr/share/container-scripts/nginx/common.sh line 17:
  local custom_dir default_dir
  ^--------------------------^ SC3043 (warning): In POSIX sh, 'local' is undefined.


In ./1.24/root/usr/share/container-scripts/nginx/common.sh line 25:
        source $custom_dir/$filename
        ^--------------------------^ SC3046 (warning): In POSIX sh, 'source' in place of '.' is undefined.
               ^-------------------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In ./1.24/root/usr/share/container-scripts/nginx/common.sh line 27:
        source $default_dir/$filename
        ^---------------------------^ SC3046 (warning): In POSIX sh, 'source' in place of '.' is undefined.
               ^--------------------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In ./1.24/root/usr/share/container-scripts/nginx/common.sh line 30:
  done <<<"$(get_matched_files "$custom_dir" "$default_dir" '*.sh' | sort -u)"
       ^-^ SC3011 (warning): In POSIX sh, here-strings are undefined.


In ./1.24/s2i/bin/run line 11:
if [ ! -v NGINX_LOG_TO_VOLUME -a -v NGINX_LOG_PATH ]; then
                              ^-- SC2166 (warning): Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.


In ./1.24/test/start-hook-test-app/nginx-start/init.sh line 1:
setup_dns_env_var() {
^-- SC2148 (error): Tips depend on target shell and yours is unknown. Add a shebang or a 'shell' directive.


In ./1.24/test/start-hook-test-app/nginx-start/init.sh line 3:
            export DNS_SERVER=`cat /etc/resolv.conf | grep "nameserver " | awk '{print $2}' | tr '\n' ' '`
                   ^--------^ SC2155 (warning): Declare and assign separately to avoid masking return values.


In ./1.22/test/start-hook-test-app/nginx-start/init.sh line 1:
setup_dns_env_var() {
^-- SC2148 (error): Tips depend on target shell and yours is unknown. Add a shebang or a 'shell' directive.


In ./1.22/test/start-hook-test-app/nginx-start/init.sh line 3:
            export DNS_SERVER=`cat /etc/resolv.conf | grep "nameserver " | awk '{print $2}' | tr '\n' ' '`
                   ^--------^ SC2155 (warning): Declare and assign separately to avoid masking return values.


In ./1.20/test/start-hook-test-app/nginx-start/init.sh line 1:
setup_dns_env_var() {
^-- SC2148 (error): Tips depend on target shell and yours is unknown. Add a shebang or a 'shell' directive.


In ./1.20/test/start-hook-test-app/nginx-start/init.sh line 3:
            export DNS_SERVER=`cat /etc/resolv.conf | grep "nameserver " | awk '{print $2}' | tr '\n' ' '`
                   ^--------^ SC2155 (warning): Declare and assign separately to avoid masking return values.

For more information:
  https://www.shellcheck.net/wiki/SC2145 -- Argument mixes string and array. ...
  https://www.shellcheck.net/wiki/SC2148 -- Tips depend on target shell and y...
  https://www.shellcheck.net/wiki/SC1090 -- ShellCheck can't follow non-const...
```

Most of the defects are due to missing shebang or shellcheck directives (`# shellcheck shell=bash`).

Because I have pushed the ShellCheck workflow into master, I can see all detected defects in Security Dashboard and review them there:

Screenshot from 2023-08-04 10-04-54

**Comment by zmiklank** [comment]
@jamacku Hi, the time has come for me to get back to this issue :).
Thanks for the analysis - how is that the **differential** shellcheck found so many errors? What was the difference here? Or rather - what part of code was actually checked?

**Comment by zmiklank** [comment]
I went through the errors and warnings and came to a conclusion that it is not feasible for us to clean all sclorg repos from them.
Vast majority of them are in some sense false positives and there is no advantage in fixing them. Even when setting some global shellcheck options, like shell=bash, as @jamacku correctly suggested, the is enormous number of them. I think that the disadvantages here outweighs the advantages -> fixing the errors would require a lot of changes throughout the codebase and can lead to regressions.
However, I guess enabling the differential shellcheck can bring us some benefits - at least not introducing new bash (syntax) errors if the shellcheck runs on patches from PRs.

TLDR: I no longer think that shellchecking whole repo is useful.

**Comment by phracek** [comment]
[test]
